Added Resource Usage section to Resource Quota Details page

dtaylor113 opened the pull request:
**Administration -> Resource Quotas -> Click on a quota, at bottom of details page:**
![image](https://user-images.githubusercontent.com/12733153/49038216-4c23de00-f18a-11e8-9712-b1453daf1c45.png)

![image](https://user-images.githubusercontent.com/12733153/49038227-55ad4600-f18a-11e8-9bed-f8efdf643f33.png)

![image](https://user-images.githubusercontent.com/12733153/49038246-62319e80-f18a-11e8-84cb-938a867aff76.png)

JIRA: https://jira.coreos.com/browse/CONSOLE-685
* PR does not include guage/donut charts, these will be done in a follow up PR

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -13,6 +13,100 @@ const menuActions = [...common];
+    'services.nodeports': 'Service Node Ports',
+  };
+
+  const nameFormatMap = {
```
Better to have one map that has both lowercase and title case values. That way we can make sure the maps have all the same keys.

```js
const quotaResourceLabels = Object.freeze({
  configmaps: { label: 'config maps', titleCase: 'Config Maps' },
  // ...
});

const humanizeQuotaResource = (resourceType, useTitleCase) =>
  _.get(quotaResourceLabels, [resourceType, useTitleCase ? 'titleCase' : 'label'], resourceType);
```

dtaylor113 replied:
fixed

spadgett commented on frontend/public/components/_quota.scss:
```diff
@@ -0,0 +1,7 @@
+}
+
+.co-m-resource-quota-empty {
+  font-size: 15px;
```
Curious why we need a special rule here.

dtaylor113 replied:
Because we are mixing `fa` and `pf` icons, this is needed so that `fa` icons are center aligned with the `pf` icons

spadgett replied:
It seems like there should be a `pficon-resources-empty` so we don't have to mix in the font awesome icon for this one special case.

FYI @serenamarie125 there are

* `pficon-resources-almost-empty`
* `pficon-resources-almost-full`
* `pficon-resources-full`

But no pficon for empty

spadgett commented on frontend/public/components/_quota.scss:
```diff
@@ -0,0 +1,7 @@
+.co-m-resource-quota-icon {
```
We should drop the `-m-`.

```suggestion
.co-resource-quota-icon {
```

dtaylor113 replied:
fixed

spadgett commented on frontend/public/components/_quota.scss:
```diff
@@ -0,0 +1,7 @@
+  padding-left: 34px;
+}
+
+.co-m-resource-quota-empty {
```
```suggestion
.co-resource-quota-empty {
```

dtaylor113 replied:
fixed

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
+  } else if (capacity > 1) {
+    capacityIconClass = 'pficon-warning-triangle-o';
+  }
+  const classPrefix = capacityIconClass.startsWith('pficon') ? 'pficon' : 'fa';
```
Let's just add `fa` or `pficon` to the strings above rather than trying to calculate a prefix. The icons are all known, and it's less error prone.

dtaylor113 replied:
I can't win :-) I initially had it that way, but thought you'd prefer the more concise logic :-)

dtaylor113 replied:
fixed

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
+    let unitType = units.getUnitType(max);
+    if (!unitType) {
+      unitType = units.getUnitType(used);
+    }
```
```js
const unitType = units.getUnitType(max) || units.getUnitType(used);
```

dtaylor113 replied:
fixed

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
+  return <i className={`${classPrefix} ${capacityIconClass}`} aria-hidden="true" />;
+};
+
+const getCapacity = (used, max, resourceType) => {
```
This seems misnamed. Capacity usually means the total amount available. Maybe `getDecimalUsage` here and `UsageIcon` above?

dtaylor113 replied:
fixed

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
+    if (!unitType) {
+      unitType = units.getUnitType(used);
+    }
+    // max cpu values do not have units, convert number and add unit for capacity calculation
```
This is wrong. Max CPU values CAN have a unit that we need to account for. We need to make sure we're using the utilties from units for any compute resource.

dtaylor113 replied:
If there is a unit for max cpu it will be used as is, this only handles when there is not a unit.

spadgett replied:
But min could also not have a unit. There's nothing special about max CPU. It's also possible to specify memory values in bytes with no units.

dtaylor113 replied:
I am seeing examples where min has a unit, but max does not:

![image](https://user-images.githubusercontent.com/12733153/48725048-f8008300-ebf8-11e8-9a42-da4f491cc17e.png)
![image](https://user-images.githubusercontent.com/12733153/48725103-21211380-ebf9-11e8-9e49-4b486c43ce5d.png)

spadgett replied:
Right, but that's just one specific example. It's not required for any CPU or memory value. `requests.cpu` could be "1" and `limits.cpu` could be 1500m.

You should just be able to always call `dehumanize` and work with the number that comes back.

spadgett replied:
> Well, the `200m` will result in a unitType of 'numeric', but calling `units.dehumanize('1', 'numeric')` results in a 0 -the logic in dehumanize doesn't handle this case.

You need to pass the CPU type for it to handle it correctly. It's not numeric. It's CPU.

I don't think you can infer the unit type. You need to specifically pass CPU for CPU, memory for memory, etc.

dtaylor113 replied:
I do not see a CPU type in  https://github.com/openshift/console/blob/master/frontend/public/components/utils/units.js#L6  ?

dtaylor113 replied:
The two CPU dehumanize examples I see are using 'numeric': https://github.com/openshift/console/blob/master/frontend/public/components/pod.jsx#L161
https://github.com/openshift/console/blob/master/frontend/public/components/pod.jsx#L165

spadgett replied:
> The two CPU dehumanize examples I see are using 'numeric': 

OK, I see now. I think that's just a bug. We need to add correct handling for CPU units.

dtaylor113 replied:
> OK, I see now. I think that's just a bug. We need to add correct handling for CPU units.

Ok, added `dehumanizeCPU` method

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -13,6 +13,100 @@ const menuActions = [...common];
+};
+
+const getResourceTypes = (quota) => {
+  const specHard = _.get(quota, 'spec.quota.hard') || _.get(quota, 'spec.hard');
```
Are you trying to handle ClusterResourceQuota? Let's do that in a separate PR.

dtaylor113 replied:
fixed

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
+};
+
+const ResourceUsageRow = ({quota, resourceType}) => {
+  const max = quota.status.hard ? quota.status.hard[resourceType] : quota.spec.hard[resourceType];
```
```suggestion
  const max = _.get(quota, ['status', 'hard', resourceType]) || _.get(quota, ['spec', 'hard', resourceType]);
```

(And similar for below.)

dtaylor113 replied:
fixed

spadgett commented on frontend/public/components/utils/units.js:
```diff
@@ -147,6 +147,78 @@ units.dehumanize = (value, typeName) => {
   return convertValueWithUnitsToBaseValue(value, type.units, type.divisor);
 };
 
+const humanizeUsageUnit = (unit, type, singular) => {
```
Why can't we use the existing humanize methods? There is already `units.humanize`

spadgett replied:
Actually, I'd suggest we should just show the string value as-is since that's what we do in other places in the console. Then we can use `units.dehumanize` for comparing values and calculating percent usage.

spadgett replied:
> Actually, I'd suggest we should just show the string value as-is since that's what we do in other places in the console.

For example,

![okd 2018-11-16 14-02-38](https://user-images.githubusercontent.com/1167259/48641624-5e429700-e9a8-11e8-84dd-7eb7fba864b6.png)

We should show these the same way here to avoid confusion.

dtaylor113 replied:
@spadgett, ok with showing units 'as-is', but wondering about the CPU units of 'cores'.  Max CPU only shows an integer (ex. '1') should this be converted to `1000 m'? -thanks

spadgett replied:
I would leave it as-is for now at least.

dtaylor113 replied:
fixed

spadgett commented on frontend/public/components/utils/units.js:
```diff
@@ -147,6 +147,78 @@ units.dehumanize = (value, typeName) => {
+  }
+};
+
+const usageAmountAndUnit = (value, type, humanizeUnits) => {
```
You shouldn't need this. Use `units.dehumanize`.

dtaylor113 replied:
fixed

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
+          <div className="col-xs-3">Max</div>
+        </div>
+        <div className="co-m-table-grid__body">
+          {getResourceTypes(rq).map((t, i) => <ResourceUsageRow key={i} quota={rq} resourceType={t} />)}
```
```suggestion
          {getResourceTypes(rq).map(type => <ResourceUsageRow key={type} quota={rq} resourceType={type} />)}
```

dtaylor113 replied:
fixed

spadgett commented:
@cshinn @dtaylor113 I like the way this is presented 👍

dtaylor113 commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -13,6 +13,100 @@ const menuActions = [...common];
+    return resourceType;
+  }
+
+  const nameTitleCaseFormatMap = {
```
I see usage like:
```
import { Map as ImmutableMap } from 'immutable';
export const resourceDetailPages = ImmutableMap<GroupVersionKind | string, () => Promise<React.ComponentType<any>>>()
  .set(...
```
However, I'm in a `.jsx` file whereas 'immutable' seems to be a Typescript library.  Ok if I use `let map = new Map();`  I don't think it's immutable though.

spadgett replied:
You shouldn't need TypeScript.

https://facebook.github.io/immutable-js/docs/#/

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -13,6 +13,100 @@ const menuActions = [...common];
+  }
+
+  // Memory Request
+  if (left === 'memory' || left === 'requests.memory') {
```
I'd suggest we order it like

CPU (requests)
CPU (limits)
Memory (requests)
Memory (limits)

dtaylor113 commented:
/retest

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -85,16 +85,15 @@ const UsageIcon = ({capacity}) => {
-    if (_.includes(resourceType, '.cpu') && unitType && !_.isNaN(maxNum)) { // skip if max cpu already has a unit
-      maxStr = `${maxNum * 1000}m`;
+    let usedNum = 0, maxNum = 0;
+    if (_.includes(resourceType, '.cpu')) {
```
I'm a little uneasy about the `getUnitType` detection. I'm not sure we can reliably detect.

I suggest creating sets of known CPU and memory/size quota resources and calling the right dehumanize method for those.

spadgett replied:
You should be able to use #852 to get the base values now without needing to check unit types.

spadgett commented on frontend/public/components/utils/units.js:
```diff
@@ -138,15 +138,58 @@ const humanize = units.humanize = (value, typeName, useRound = false) => {
 export const humanizeNumber = v => humanize(v, 'numeric', true).string;
+export const humanizeCPU = v => humanize(v, 'numeric', true).string;
+export const dehumanizeCPU = v => {
+  const _dehumanizeCPU = _.spread((value, unit) => {
```
Prefer es6 syntax to `_.spread`:

```suggestion
  const _dehumanizeCPU = ([value, unit]) => {
```

spadgett commented on frontend/public/components/utils/units.js:
```diff
@@ -138,15 +138,58 @@ const humanize = units.humanize = (value, typeName, useRound = false) => {
+    if (!unit) {
+      return {value: value * 1000, unit: 'm'};
+    }
+    return {value, unit};
```
This is not guaranteed to return values that always have the same unit.

dtaylor113 replied:
for CPU isn't it 'm' or nothing?  What other units could there be?

dtaylor113 commented:
> The way you've written this, I think we can reuse the existing dehumanize logic with type:
> 
> ```js
> cpu: {
>   units: ['m', ''],
>   space: false,
>   divisor: 1000,
> },
> ```
> 
> Can you try this first?

@spadgett 

const test = units.dehumanize('1', 'cpu').value;     // returns "1"
const test2 = units.dehumanize('100m', 'cpu').value;  // returns "100m"

* trying other values in units array, so far no luck

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -13,6 +13,42 @@ const menuActions = [...common];
+  'secrets': {label: 'secrets', titleCase: 'Secrets'},
+  'services': {label: 'services', titleCase: 'Services'},
+  'services.loadbalancers': {label: 'service load balancers', titleCase: 'Service Load Balancers'},
+  'services.nodeports': {label: 'service node ports', titleCase: 'Service Node Ports'},
```
@dtaylor113 This looks like an incomplete list.

https://docs.okd.io/latest/admin_guide/quota.html#managed-by-quota

I'm worried that this is constantly going to get out of date, so we have a mix of humanized / unhumanized labels. We might skip this and just show the real value. It's also confusing to see something different in the table than you need to add the to YAML.

cc @cshinn

dtaylor113 replied:
@spadgett, @cshinn, so instead of:
![image](https://user-images.githubusercontent.com/12733153/49018834-6e056c80-f15a-11e8-9760-e0fd0ee8f063.png)
We'd show:
```
    limits.cpu
    requests.cpu
    limits.memory
    requests.memory
    pods
```

spadgett replied:
@dtaylor113 I think that's better because we're never going to be able to humanize everything, and things will always be added later. Storage classes follow the format `<storage-class-name>.storageclass.storage.k8s.io/requests.storage`, so those will pretty much always have the raw name unless we do some regex matching or something. Plus it doesn't match what's in the YAML or CLI or docs, so it's harder to know what to change if you want to edit something. We had a number of bugs around trying to humanize quota keys in the previous console.

I think this is a losing battle, and we should just show the quota key as-is.

dtaylor113 replied:
Ok, thanks, sounds reasonable to me.  @cshinn, sound good to you?  -thanks

cshinn replied:
yeah, that works for me. Especially if we're likely to have other things added

dtaylor113 commented:
Hi @spadgett, updated to use new 'convertToBaseValue` function.

dtaylor113 commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -30,11 +66,65 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
+  } else if (capacity > 1) {
+    capacityIconClass = 'pficon pficon-warning-triangle-o';
+  }
+  return <i className={`${capacityIconClass}`} aria-hidden="true" />;
```
fixed

dtaylor113 commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -13,6 +13,42 @@ const menuActions = [...common];
+const compareResourceType = (left, right) => {
+  left = humanizeQuotaResource(left);
+  right = humanizeQuotaResource(right);
+  return left.localeCompare(right);
```
done.  new sort implemented

dtaylor113 commented:
I believe I have addressed all review comments.  Pls see this PR's updated Description of screenshots of the usage table with unhuman labels and new sort.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -13,6 +13,18 @@ const menuActions = [...common];
 
+const compareResourceType = (left, right) => {
+  const regex = /\.|\//; //find object types delimited by '.' or '/'. Ex: '.cpu', '.memory', or '/persistentvolumeclaims'
+  left = left.match(regex) ? left.split(regex).reverse().join('') : left;
```
You shouldn't need to check for matches. `split` will return an array with one value (the string) when there are no matches.

```suggestion
  left = left.split(regex).reverse().join('');
```

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -13,6 +13,18 @@ const menuActions = [...common];
 const quotaKind = quota => quota.metadata.namespace ? referenceForModel(ResourceQuotaModel) : referenceForModel(ClusterResourceQuotaModel);
 
+const compareResourceType = (left, right) => {
+  const regex = /\.|\//; //find object types delimited by '.' or '/'. Ex: '.cpu', '.memory', or '/persistentvolumeclaims'
```
To avoid repeating the split/reverse/join logic, I'd do something like

```js
// Reverse segments around `.` and `/` delimiters.
const reverseSegments = str => str.split(/\.|\//).reverse.join('');
left = reverseSegments(left);
right = reverseSegments(right);
return left.localCompare(right);
```

I wouldn't worry about case insensitive sorts if we're not sorting humanized values.

spadgett commented:
I pushed a small fix for the last comment
